Tesla boost for RA2 mod

matija-hustic opened the pull request:
Functionality needed for the RA2 mod. Also described in #2330

Introduces a new upgrade called `tesla`. Its consumers would be
- `RequiresPower` that would disable itself (thus not necessitating power for proper functioning of the actor) when enough `tesla` upgrades  have been granted
- `Armament` that would increase range, damage and reload speed and modify rendering for a juiced-up tesla coil

Normally tesla coils would use these traits.

The upgrade's provider would be the `TeslaBoost` activity which requires the actor, that wishes to use this activity, to have the `GivesTeslaBoost` trait. Normally, tesla tanks and tesla troopers would use this trait.

I've tried to make things flexible by allowing
- the modders to decide whether or not `GivesTeslaBoost` activates the `TeslaBoost` when the units are idle in the vicinity of a receiving tesla coil
- the player to force the tesla boost by explicitly giving an order to a tesla tank or tesla trooper to target a particular tesla coil
- the `TeslaBoost` activity to be used with mobile units

![untitled](https://cloud.githubusercontent.com/assets/5747153/8380918/78835298-1c20-11e5-8a50-dbe2d0632432.png)

matija-hustic committed:
Added tesla boost capabilities.

Shock troopers and tesla tanks standing near a tesla coil
will charge it up, enhancing its range, modifying the zap
animation and allowing it to work even when player's power
is down.
The sound for charger weapons edited from
https://www.freesound.org/people/JoelAudio/sounds/136542/

Unit158 commented:
:-1: This should be with the RA2 mod.

On 2015-06-26, at 11:24 AM, Matija Hustić <notifications@github.com> wrote:

> Functionality needed for the RA2 mod. Also described in #2330
> 
> Introduces a new upgrade called tesla. Its consumers would be
> 
> RequiresPower that would disable itself (thus not necessitating power for proper functioning of the actor) when enough tesla upgrades have been granted
> Armament that would increase range, damage and reload speed and modify rendering for a juiced-up tesla coil Normally tesla coils would use these traits.
> The upgrade's provider would be the TeslaBoost activity which requires the actor, that wishes to use this activity, to have the GivesTeslaBoost trait. Normally, tesla tanks and tesla troopers would use this trait.
> 
> I've tried to make things flexible by allowing
> 
> the modders to decide whether or not GivesTeslaBoost activates the TeslaBoost when the units are idle in the vicinity of a receiving tesla coil
> the player to force the tesla boost by explicitly giving an order to a tesla tank or tesla trooper to target a particular tesla coil
> the TeslaBoost activity to be used with mobile units
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/OpenRA/OpenRA/pull/8563
> 
> Commit Summary
> 
> Tesla boost.
> File Changes
> 
> M OpenRA.Mods.Common/Traits/Power/RequiresPower.cs (11)
> A OpenRA.Mods.RA/Activities/TeslaBoost.cs (139)
> M OpenRA.Mods.RA/OpenRA.Mods.RA.csproj (2)
> A OpenRA.Mods.RA/Traits/GivesTeslaBoost.cs (198)
> A mods/ra/bits/litningupg.shp (0)  
> M mods/ra/rules/infantry.yaml (3)   
> M mods/ra/rules/structures.yaml     (15)
> M mods/ra/rules/vehicles.yaml (2)
> M mods/ra/sequences/misc.yaml     (7)
> M mods/ra/weapons/other.yaml (15)
> Patch Links:
> 
> https://github.com/OpenRA/OpenRA/pull/8563.patch
> https://github.com/OpenRA/OpenRA/pull/8563.diff
> —
> Reply to this email directly or view it on GitHub.
>

matija-hustic commented:
Ah they don't share common code here?

GraionDilach commented:
The RA2 mod currently does not have it's own DLL repo - @Phrohdoh plans to maintain one on his profile, but tbh I'm not a fan of scattering DLL repos.

Besides that, we'll need the premature upgrade dictionary initialization to fix #8444.

matija-hustic commented:
I thought that RA2 mod on @Phrohdoh's profile was only there for art copyright reasons, but that all the code resides in this repository.

GraionDilach commented:
I haven't heard a concrete word on that yet and tbh yes, I'd welcome the DLL code residing in this repository too.

GraionDilach commented:
http://logs.openra.net/?year=2015&month=06&day=26#16:03:55

pchote commented:
This has been a long-planned feature for our RA mod, so the code is in the right place.

matija-hustic commented:
I'll see what I can do with @GraionDilach's suggestions. In the meantime, this should be marked as having changes requested.

atlimit8 commented on OpenRA.Mods.RA/Activities/TeslaBoost.cs:
```diff
@@ -0,0 +1,138 @@
+					upgraded = false;
+					if (um != null)
+						foreach (var upgrade in upgrades)
+							for (var u = 0; u < grantMultiple; u++)
```
If `grantMultiple` is to be used as a cap, the it should have a name like `maxGrantCount` instead of one that implies boolean value.

atlimit8 commented on OpenRA.Mods.RA/Traits/GivesTeslaBoost.cs:
```diff
@@ -0,0 +1,196 @@
+		public readonly string[] Upgrades = { "tesla" };
+
+		[Desc("Make this actor grant the upgrades multiple times.")]
+		public readonly int GrantMultiple = 1;
```
As mentioned above, this looks like it is interpreted as a predicate value.
`"Maximum number of upgrades to grant."`
`MaxGrantCount`

matija-hustic replied:
This guy is here to make Tesla tanks grant `tesla` multiple times because if I remember correctly from RA2, one tank should give more of a boost than a single trooper

atlimit8 replied:
Regardless of its existence, its name is misleading.

matija-hustic replied:
I'm open to suggestions ;)

GraionDilach replied:
Tesla Tanks couldn't charge coils in RA2. The original code hardcoded the charger to be an infantry.

matija-hustic replied:
Are you sure about this?

GraionDilach replied:
https://github.com/Ares-Developers/YRpp/blob/master/BuildingClass.h#L230

matija-hustic replied:
Ah ok. I must be getting senile :D

atlimit8 commented:
Since this could be used for supression attacks and possibly Chrono Legionnaire attacks, perhaps `GivesTeslaBoost` should be `SupportsUpgrade` or `SupportUpgradeAttack` and `TeslaBoost` should be `SupportUpgrade` or `SupportUpgradeActivity`.

matija-hustic commented:
I'll see how much of these even stay when I try to do this using upgrade granting warheads. But `GrantMultiple` functionality will probably have to stay. So a preferred name would be appreciated

atlimit8 commented:
Is there a reason this can not be done with the existing code?

Perhaps:
  * `GrantMultiple` => `GrantCount`
  * `GivesTeslaBoost` => `GrantsUpgradeDuringAttack`
  * `TeslaBoost`  =>`UpgradeDuringAttack`

atlimit8 commented:
People on the IRC channel question, perhaps the reason should by appended to the description.

matija-hustic commented:
Upgrade granting warheads is existing code, so this would make this pr smaller. Plus it's more logical.

See discussion in

http://logs.openra.net/?year=2015&month=06&day=26#16:03:55

matija-hustic commented:
I asked GraionDilach to write the reasons here, but we thought a link to the discussion would suffice

atlimit8 commented:
Do you plan to merge it with the current GrantUpgradeWarhead?

matija-hustic commented:
Don't know yet, didn't have the time to look into it in detail. I'll do my best to do it elegantly, if possible without modifying existing functionality

GraionDilach commented:
@atlimit8: Chrono Legionnaire should not use this at all. Chrono Legionnaire should have two armaments, with one only used for upgrading the target's TargetType from casual to Temporal only along with DisableUpgrade recoloring the target to light white, a WithDecoration giving the temporal sparkles and the another one only affecting Temporal TargetType'd targets, maintaining this upgrade and delivering the damage entirely. It has no relation to this atm.... the Paratroopers refactor is more related due to the current TargetType-upgrade proposal disallowing to have different TargetType setup for paratroopers and for temporal weapons (and for RA2 IronCurtain because in RA2, AI units didn't even wasted shots for Ironcurtained units, something OpenRA does).

Both the TeslaBoost and GivesTeslaBoost trait can be replaced via an upgradable Armament on the coil with a weapon having GrantUpgrade warheads to get it upgraded on SHOK's side, with both of SHOK's armaments sharing the AmmoPool to prevent SHOK charging both the coil and attacking at the same time. This needs no change in the code at GrantUpgrade, only YAML level modding - Armaments needs a change just because of #8444 alone.

The sole thing which shall stay from the current state of this PR is the Upgradable implementation of RequiresPower. Everything else can be axed and if @matija-hustic does not have time, I gladly fill a PR against his PR tomorrow to do this YAML modding.

matija-hustic commented:
Go ahead and take care of this @GraionDilach

GraionDilach commented:
Did a PR against your PR. That should solve 90% of the nits everybody can have against the implementation.

GraionDilach commented on mods/ra/rules/infantry.yaml:
```diff
@@ -634,7 +635,7 @@ Ant:
 		Weapon: mandible
 	TargetableUnit:
-		TargetTypes: Ground, Infantry
+		TargetTypes: Ground, Infantry, TeslaBoost
```
I didn't touched this in my PR - why is this change?

matija-hustic commented:
It seems that timed upgrades cannot go over level 1. This means that without changing this behavior, the tesla powerup cannot be implemented using upgrade granting warheads...
Apart from that, `AttackBase` only handles `DamageWarheads`. This means that a new trait would still be required to control the tesla powerup `GrantUpgradeWarhead`. It would nevertheless be more elegant with the warheads, but... Maybe see if the timed upgrades could be changed in a separate PR?
In the meantime, what should I do with this PR?

penev92 commented:
I started on a warhead refactoring quest a while back (see 7997), and I still need to do one more PR. That may handle the `GrantUpgradeWarhead` side of the problem (or at least pave the road for it).
The `TimedUpgrade`s can be refactored to have levels (which would most likely want its own PR).

If you are hopelessly stuck on those two points, I would suggest closing this PR until they are resolved.

GraionDilach commented:
I honestly can't find any reason in the code why this wouldn't work.

matija-hustic commented:
I've tried doing this thing locally using higher level timed upgrades and `GrantUpgradeWarhead`.

When the upgrade granting warhead's weapon's reload time is less than the upgrade duration, it's not the prettiest of sights when
* you put one tesla trooper next to a coil and then have the coil's range or
* you put two tesla troopers next to a coil without power

![tesla1](https://cloud.githubusercontent.com/assets/5747153/8509826/f4b38af0-22b4-11e5-87e1-f41ab1c91fda.gif)
![tesla2](https://cloud.githubusercontent.com/assets/5747153/8509835/38377aca-22b5-11e5-86cb-2ed4ea8c74a1.gif)

If you put the upgrade granting warhead's weapon's reload and the upgrade durationto be exactly equal, the situation is visually nicer, but risks breaking certain behaviors due to race conditions... One example is ordering the coil to attack ground - for a certain time it does but after a random number of zaps it just cancels the `ChargeAttack` activity because the coil's upgraded armament briefly becomes disabled.

matija-hustic commented:
Idea: leave how the timed upgrade granting works right now, just make the timed upgrades source dependent (one source can grant at most one upgrade level, if it grants a new one before the old one is revoked, the old one's duration is just refreshed).

GraionDilach commented:
I think that jumpy behaviour is acceptable tbh - but if you want to add that GrantUpgadeWarhead change, do it as a variable which modders can disable.

matija-hustic commented:
Depends on #8658.

GraionDilach commented:
#8658 has been merged.

I also made a new litning SHP which might look better, @matija-hustic, if you're interested, I add it to my PR against yours.
![openra-2015-07-07t200145z](https://cloud.githubusercontent.com/assets/1136302/8721503/5d505840-2bbb-11e5-8f2d-b4f2ca573bd5.png)

matija-hustic commented:
OK, so here's the version using upgrade granting warheads.

I was unable to take advantage of `AttackBase` and its subclasses since at most one of those per actor is allowed. And this means that it is effectively impossible to use `AttackBase` to implement a powerup weapon with a different cursor from the attack cursor used for other weapons.

This is why I've stuck to `GivesTeslaBoost`.

Also, for more natural behavior (when putting idle troopers next to a coil, they'll power it up, but stop powering up and attack as soon as enemies are present; on the other hand, when issuing an explicit order to power up a coil, the trooper will not stop powering up even if enemies approach), I've also kept `TeslaBoost`.

Maybe somehow modifying `AttackBase` could eliminate the need for these two classes?

matija-hustic commented:
@GraionDilach, sure, could you please add your artwork so that I can merge it?

penev92 commented:
Well maybe we're at the point where we should have `AttackBase` upgradable as well?
You should open an issue for what you said above about it.

GraionDilach commented:
The main issue isn't with `AttackBase` but `AttackMedic` and `AutoHeal` duplicating the entire Attack* logic for themselves.

abcdefg30 commented on OpenRA.Mods.RA/Traits/GivesTeslaBoost.cs:
```diff
@@ -0,0 +1,200 @@
+	public class GivesTeslaBoost : INotifyAddedToWorld, INotifyRemovedFromWorld, ITick, IIssueOrder, IResolveOrder, IOrderVoice, INotifyIdle
+	{
+		readonly string orderName = "TeslaBoost";
+		readonly System.Drawing.Color targetLineColor = System.Drawing.Color.BlueViolet;
```
Why not `using System.Drawing;` and then drop it here?

abcdefg30 commented on OpenRA.Mods.RA/Traits/GivesTeslaBoost.cs:
```diff
@@ -0,0 +1,200 @@
+			this.info = info;
+
+			mobile = self.Trait<Mobile>();
+			armament = self.TraitsImplementing<Armament>().Where(a => a.Info.Name == info.Armament).First();
```
`self.TraitsImplementing<Armament>().First(a => a.Info.Name == info.Armament);`?

abcdefg30 commented on OpenRA.Mods.RA/Activities/TeslaBoost.cs:
```diff
@@ -0,0 +1,58 @@
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.GameRules;
+using OpenRA.Mods.Common.Activities;
```
This using and `OpenRA.GameRules` aren't used.

matija-hustic commented:
@abcdefg30, thanks for reviewing, but this PR depends on #8768. With it, `TeslaBoost.cs` and `GivesTeslaBoost.cs` will become unnecessary. This PR should be marked as having dependencies...

abcdefg30 commented:
Oh, heh. Added the label.

Unit158 commented:
This will need a rebase

matija-hustic commented:
So finally #8768 was merged and this PR is updated. I still need to test that everything works well with autotargeting and that nothing has regressed.
Special thanks to @GraionDilach for helping me out a lot!

pchote commented on mods/ra/weapons/other.yaml:
```diff
@@ -177,6 +208,22 @@ TTankZap:
+	Warhead@charge: GrantUpgrade
+		Range: 42
+		Duration: 125
+		Upgrades: tesla, tesla1, tesla2
```
Why do you have three different upgrade types that all do the same thing?

matija-hustic replied:
Because of timed upgrades again...
Only one levelup is permitted per timed upgrade source. This is OK for shock troopers since they only give one tesla up each, but it is a little problematic for tesla tanks since they need to give three tesla ups (obviously all with the same source).

Simply allowing more than one levelup per timed upgrade source is kinda problematic - take a shock trooper that grants a 75 tick upgrade every 70 ticks. If you allow multiple upgrades per source, then for a brief period of 5 ticks, the shock trooper would give two tesla ups to the tesla coil. If you make the shock trooper grant a 70 tick upgrade every 70 ticks, then the upgrade level can at times actually drop to 0 on the 70th tick (which causes e.g. that the tesla coil loses its target and other weird stuff).

But yeah, this isn't too pretty. Maybe it'd be nicer if tesla tanks gave a different kind of upgrade, that would count as three shock trooper upgrades?

pchote replied:
How difficult would it be to add an `UpgradeLevels:` parameter that we set to 3 for the tesla tank?

matija-hustic replied:
I've made the `UpgradeManager` support requests to grant a timed upgrade multiple times from a single source. The maximum number of times it will grant needs to be supplied by the caller (in this instance `GrantUpgradeWarhead`), otherwise it is assumed to be 1.

What `GrantUpgradeWarhead` then does when it gets `Upgrades == { "tesla", "tesla", "tesla" }`, it requests granting of each of these tesla ups, but notifying that it should be granted up to three times from this particular source.

How does this sound?

matija-hustic committed:
Enabled multiple timed upgrades from single source.

This makes the UpgradeManager support requests to grant a timed
upgrade multiple times from a single source.
GrantUpgradeWarhead modified to take advantage of this.

pchote commented:
Code looks good now, and this is feels really well integrated with the game :+1:

The only possible downside is that this may make the shock troopers even more OP, but we can deal with that balance issue during playtesting.

obrakmann commented:
Looks fine to me. But yeah, we should probably nerf the shock troopers some more, since they've now gained a defensive capability as well. I'd suggest increasing the price and lowering the damage somewhat. But that's for another PR. :+1:

obrakmann commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/be9347a9f8426d6a597ff31645e7be0dffd5a797)
